pennylane/devices/qubit/apply_operation.py

```diff
@@ -160,50 +181,43 @@ def _(op: type_op, state):
         [1., 0.]], requires_grad=True)
 
     """
-    if (len(op.wires) < EINSUM_OP_WIRECOUNT_PERF_THRESHOLD) and (
-        math.ndim(state) < EINSUM_STATE_WIRECOUNT_PERF_THRESHOLD
-    ):
-        return apply_operation_einsum(op, state)
-    return apply_operation_tensordot(op, state)
+    if (
+        len(op.wires) < EINSUM_OP_WIRECOUNT_PERF_THRESHOLD
+        and math.ndim(state) < EINSUM_STATE_WIRECOUNT_PERF_THRESHOLD
+    ) or (op.batch_size and batch_dim):
+        return apply_operation_einsum(op, state, batch_dim=batch_dim)
+    return apply_operation_tensordot(op, state, batch_dim=batch_dim)
 
 
 @apply_operation.register
-def apply_paulix(op: qml.PauliX, state):
+def apply_paulix(op: qml.PauliX, state, batch_dim=0):
     """Apply :class:`pennylane.PauliX` operator to the quantum state"""
-    return math.roll(state, 1, op.wires[0])
+    axis = op.wires[0] + batch_dim
+    return math.roll(state, 1, axis)
 
 
 @apply_operation.register
-def apply_pauliz(op: qml.PauliZ, state):
+def apply_pauliz(op: qml.PauliZ, state, batch_dim=0):
     """Apply pauliz to state."""
-    n_wires = math.ndim(state)
-    sl_0 = _get_slice(0, op.wires[0], n_wires)
-    sl_1 = _get_slice(1, op.wires[0], n_wires)
-
-    state1 = math.multiply(state[sl_1], -1)
-    return math.stack([state[sl_0], state1], axis=op.wires[0])
-
-
-@apply_operation.register
-def apply_phase(op: qml.PhaseShift, state):
```

Comment:
why are you deleting this?

Reply:
It does not work with batching if the operator is batched and the state isn't, or when both the state and operation are batched.

Fixing the dispatched function would involve either looping through the batched parameters/states or some variation of tensordot/einsum, at which point it doesn't really make sense to even have the dispatched function.

Reply:
gotcha, thanks for the details! if the performance difference is significant, I think we might want to tolerate the zig-zag decision-making to support this custom apply_op implementation. If it's too ugly for no good reason, then I suppose I could be ok to delete this. It just kinda feels like wasted knowledge

Reply:
Might be good to check back with @albi3ro and/or performance squad? Are there performance considerations with this device at this level already? :)

Reply:
I'm fine with deleting this for now.  I had to delete a lot of other "special implementation" methods because they turned out to be slower. As long as we keep around the special implementation for something important like `PauliZ`, I'm less concerned with `PhaseShift`.